modules/core/shared/src/main/scala/io/circe/Decoder.scala

```diff
@@ -612,6 +612,15 @@ object Decoder
     }
   }
 
+  /**
+   * @group Decoding
+   */
+  implicit final val decodeNothing: Decoder[Nothing] = new Decoder[Nothing] {
+    final def apply(c: HCursor): Result[Nothing] = c.value match {
```

Comment:
You can avoid some syntactic / runtime overhead here by just ignoring the cursor and returning the `Left` directly.